File: Library.py
from Books import Books


class Library:
    def __init__(self):
        self.availableBooks = []
        self.users = []

    def showActiveUsers(self):
        if len(self.users) == 0:
            print("There are no registered books")
        else:
            print("These users own at least one rented book.")
            for user in self.users:
                if user.getRentCounter() > 0:
                    print("|User Name: |" + user.getName() 
                        + "| User Last Name: |" + user.getLastName() 
                        + "| User Age: |" + user.getAge() 
                        + "| User ID: |" + user.userID())

    def activeBooks(self):
        if len(self.availableBooks) == 0:
            print("There are no registered books")
        else:
            for book in self.availableBooks:
                if book.getRentCounter() > 0:
                    print("| TITLE: |" + book.getTitle() + "| NUMBER OF RENTS: | " + book.getRentCounter())
                    if not book.isRented():
                        print(" -Without Owner- ")
                    else:
                        print(" | OWNER: | " + book.getOwner().getName())

    def inactiveBooks(self):
        if len(self.availableBooks) == 0:
            print("There are no registered books")
        else:
            print("These books have not been rented")
            for book in self.availableBooks:
                if book.getRentCounter() == 0:
                    print("| TITLE: |" + book.getTitle() + "| RENT NUMBER: | " + book.getRentCounter())

    def addUser(self, user):
        self.users.append(user)

    def showAllUsers(self):
        for user in self.users:
            print("-------------------------------")
            print("|User Name: |" + user.getName() 
                    + "| User Last Name: |" + user.getLastName() 
                    + "| User Age: |" + user.getAge() 
                    + "| User ID: |" + user.userID())
            print("-------------------------------")
            if len(user.getRentedBooks()) == 0:
                print("This user has not rented any books")
            else:
                print("_____________Rented Books_____________")
                for book in user.getRentedBooks():
                    print("| TITULO: |", book.getTitle())

    def rentBook(self):
        if len(self.users) == 0 or len(self.availableBooks) == 0:
            print("No users or books to rent")
        else:
            i = self.getBookIndex()
            if self.validBookIndex(i):
                if not self.availableBooks[i].isRented():
                    j = self.getUserIndex()
                    if self.validUserIndex(j):
                        self.availableBooks[i].rentBook(self.users[j])
                        self.users[j].rentBook(self.availableBooks[i])
                    else:
                        print("User not found")
                else:
                    print("Already rented by: " + self.availableBooks[i].getOwner().getName())
            else:
                print("Book not found")

    def addBook(self):
        print("Enter the title of the book to add: ")
        title = input()
        book = Books(title)
        self.availableBooks.append(book)
        print("Book added succesfully.")

    def returnBook(self):
        if len(self.availableBooks) == 0:
            print("There are no registered books")
        else:
            i = self.getUserIndex()
            if self.validUserIndex(i):
                if len(self.users[i].getRentedBooks()) != 0:
                    n = self.getBookIndex(i)
                    if self.validBookIndex(n):
                        self.users[i].returnBook(self.availableBooks[n])
                        self.availableBooks[n].returnBook()
                        print("The book has been returned to the bookstore")
                    else:
                        print("The book was not found")
                else:
                    print("This user has not rented anything")
            else:
                print("User not found")

    def showUserNames(self):
        print("These are the available IDs")
        print("----------------------------------")
        for user in self.users:
            print(" | NAME: |" + user.getName() + " | ID: |" + user.userID())

    def getUserIndex(self):
        verification = -1
        self.showUserNames()
        print("Enter the user ID: ")
        ID = input()
        for i in range(len(self.users)):
            if ID == self.users[i].userID():
                verification = i
                break
        return verification

    def validUserIndex(self, valid):
        if valid == -1:
            return False
        else:
            return True

    def showBookTitles(self):
        print("These are the available IDs")
        print("----------------------------------")
        for book in self.availableBooks:
            print(" | TITLE: | " + book.getTitle() + " | Book ID: " + book.getID())

    def showBookTitles(self, i):
        print("These are the IDs available to the user: %s", self.users[i].getName())
        print("----------------------------------")
        for book in self.users[i].getRentedBooks():
            print(" | TITLE: | " + book.getTitle() + " | Book ID: " + book.getID())

    def getBookIndex(self):
        verification = -1
        self.showBookTitles()
        print("Enter the book ID")
        ID = input()
        for i in range(len(self.availableBooks)):
            if ID == self.availableBooks[i].getID():
                verification = i
                break
        return verification

    def getBookIndex(self, i):
        verification = -1
        self.showBookTitles(i)
        print("Enter the book ID")
        ID = input()
        for j in range(len(self.availableBooks)):
            if ID == self.users[i].getRentedBooks()[j].getID():
                verification = j
                break
        return verification

    def validBookIndex(self, valid):
        if valid == -1:
            return False
        else:
            return True

    def showUserData(self):
        if len(self.users) == 0:
            print("")
        else:
            i = self.getUserIndex()
            if self.validUserIndex(i):
                print("0============= USER INFORMATION =============0")
                print("|User Name: |" + self.users[i].getName() 
                    + "| User Last Name: |" + self.users[i].getLastName() 
                    + "| User Age: |" + self.users[i].getAge() 
                    + "| User ID: |" + self.users[i].userID())
                if len(self.users[i].getRentedBooks()) == 0:
                    print("This user has not rented any books")
                else:
                    print("_____________Rented Books_____________ ")
                    for book in self.users[i].getRentedBooks():
                        print(" | TITLE: " + book.getTitle())
                        print()
                print()
            else:
                print("User not found")

    def showAvailableBooks(self):
        if len(self.availableBooks) == 0:
            print("No hay libros registrados")
        else:
            print(" ------------AVAILABLE BOOKS------------")
            for book in self.availableBooks:
                print("| TITLE: |" + book.getTitle() + "| NUMBER OF RENTS: | " +  book.getRentCounter())
                if not book.isRented():
                    print(" | Without Owner | ")
                else:
                    print(" | OWNER: | " + book.getOwner().getName())

File: Library/LIbrary/src/library/Library.java
package library;
import java.util.*;
class Library {
    private ArrayList<Books> availableBooks = new ArrayList<>();
    private ArrayList<Users> users = new ArrayList<>();
    private Scanner sc = new Scanner(System.in);
    public Library(){
        
    }
     public  void showActiveUsers(){
        if(users.isEmpty()){
            System.out.println("There are no registered books");
        }
        else{
            System.out.println("These users own at least one rented book.");
            for (Users users : users) {
                if(users.getRentCounter() > 0){
                   System.out.println("|User Name: |" + users.getName() 
                    + "| User Last Name: |" + users.getLastName() 
                    + "| User Age: |" + users.getAge() 
                    + "| User ID: |" + users.userID());
                }
            }
        }
    }
    public  void activeBooks(){
        if(availableBooks.isEmpty()){
            System.out.println("There are no registered books");
        }
        else{
            for (Books book : availableBooks) {
                if(book.getRentCounter() > 0){
                    System.out.printf("| TITLE: |" + book.getTitle() + "| NUMBER OF RENTS: | " + book.getRentCounter());
                    if(!book.isRented()){
                        System.out.print(" -Without Owner- ");
                    }
                    else{
                        System.out.printf(" | OWNER: | " + book.getOwner().getName());
                    }
                    System.out.println();
                }
            }
        }
    }

    public  void inactiveBooks(){
        if(availableBooks.isEmpty()){
            System.out.println("There are no registered books");
        }
        else{
            System.out.println("These books have not been rented");
            for (Books book : availableBooks) {
                if(book.getRentCounter() == 0){
                    System.out.printf("| TITLE: |" + book.getTitle() + "| RENT NUMBER: | " + book.getRentCounter());
                }
            }
        }
    }

     public void addUser(Users user) {
         users.add(user);
      }
    
    public void showAllUsers(){
          for (Users users : users){
            System.out.println("-------------------------------");
            System.out.println("|User Name: |" + users.getName() 
                    + "| User Last Name: |" + users.getLastName() 
                    + "| User Age: |" + users.getAge() 
                    + "| User ID: |" + users.userID());
            System.out.println("-------------------------------");;
                if(users.getRentedBooks().isEmpty()){
                    System.out.println("This user has not rented any books");
                }
                else{
                    System.out.println("_____________Rented Books_____________");
                    for (Books book : users.getRentedBooks()) {
                        System.out.printf("| TITULO: |", book.getTitle());
                    }
                }
          }
      }

    public void rentBook() {
      if(users.isEmpty() || availableBooks.isEmpty()){
            System.out.println("No users or books to rent");
        }
        else{
            int i = getBookIndex();
            if(validBookIndex(i)){
                if (!availableBooks.get(i).isRented()){
                    int j = getUserIndex();
                    if(validUserIndex(j)){
                        availableBooks.get(i).rentBook(users.get(j));
                         users.get(j).rentBook(availableBooks.get(i));
                    }
                    else{
                        System.out.println("User not found");
                    }
                }
                else{
                    System.out.println("Already rented by: " + availableBooks.get(i).getOwner().getName());
                }
            }
            else{
                System.out.println("Book not found");
            }
        }
    }
    public void addBook() {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the title of the book to add: ");
        String title = sc.nextLine();
        Books book = new Books(title);
        availableBooks.add(book);
        System.out.println("Book added succesfully.");
    }

   
    public void returnBook(){
        if(availableBooks.isEmpty()){
            System.out.println("There are no registered books");
        }
        else{
            int i = getUserIndex();
            if(validUserIndex(i)){
                if(!users.get(i).getRentedBooks().isEmpty()){
                    int n = getBookIndex(i);
                    if(validBookIndex(n)){
                        users.get(i).returnBook(availableBooks.get(n));
                        availableBooks.get(n).returnBook();
                        System.out.println("The book has been returned to the bookstore");
                    }
                    else{
                        System.out.println("The book was not found");
                    }
                }
                else{
                    System.out.println("This user has not rented anything");
                }
            }
            else{
                System.out.println("User not found");
            }
        }  
    }
     private  void showUserNames(){
        System.out.println("These are the available IDs");
        System.out.println("----------------------------------");
        for (Users user : users) {
            System.out.printf(" | NAME: |" + user.getName() + " | ID: |" + user.userID());
            System.out.println();

        }
    }
     private int getUserIndex(){
        int verification = -1;
        showUserNames();
        System.out.println("Enter the user ID: ");
        String ID = sc.next();

        for (int i = 0; i < users.size(); i++) {
            if(ID.equals(users.get(i).userID())){
                verification = i;
                break;
            }
        }
        return verification;
    }
 private boolean validUserIndex(int valid){
        if(valid == -1){
            return false;
        }
        else{
            return true;
        }
    }
  private void showBookTitles(){
        System.out.println("These are the available IDs");
        System.out.println("----------------------------------");
        for (Books book : availableBooks) {
            System.out.printf(" | TITLE: | " + book.getTitle() + " | Book ID: " + book.getID());
        }
    }
  private void showBookTitles(int i){
        System.out.printf("These are the IDs available to the user: %s", users.get(i).getName());
        System.out.println("----------------------------------");
        for (Books book : users.get(i).getRentedBooks()) {
           System.out.printf(" | TITLE: | " + book.getTitle() + " | Book ID: " + book.getID());
        }
    }

    private int getBookIndex(){
        int verification = -1;
        showBookTitles();
        System.out.println("Enter the book ID");
        String ID = sc.next();
        for (int i = 0; i < availableBooks.size(); i++) {
            if(ID.equals(availableBooks.get(i).getID())){
                verification = i;
                break;
            }
        }
        return verification;
    }

    private int getBookIndex(int i){
        int verification = -1;
        showBookTitles(i);
        System.out.println("Enter the book ID");
        String ID = sc.next();

        for (int j = 0; j < availableBooks.size(); j++) {
            if(ID.equals(users.get(i).getRentedBooks().get(j).getID())){
                verification= j;
                break;
            }
        }
        return verification;
    }

    private boolean validBookIndex(int valid){
        if(valid == -1){
            return false;
        }
        else{
            return true;
        }
    }

    public void showUserData(){
        if(users.isEmpty()){
            System.out.println("");
        }
        else{
            int i = getUserIndex();
            if(validUserIndex(i)){
               System.out.println("0============= USER INFORMATION =============0");
               System.out.println("|User Name: |" + users.get(i).getName() 
                    + "| User Last Name: |" + users.get(i).getLastName() 
                    + "| User Age: |" + users.get(i).getAge() 
                    + "| User ID: |" + users.get(i).userID());
                if(users.get(i).getRentedBooks().isEmpty()){
                   System.out.println("This user has not rented any books");
                }
                else{
                    System.out.println("_____________Rented Books_____________ ");
                    for (Books book : users.get(i).getRentedBooks()) {
                        System.out.printf(" | TITLE: " + book.getTitle());
                        System.out.println();
                    }
                }
                System.out.println();
            }
            else{
                System.out.println("User not found");
            }

        }
    }
    public void showAvailableBooks(){
       if(availableBooks.isEmpty()){
            System.out.println("No hay libros registrados");
        }
        else{
            System.out.println(" ------------AVAILABLE BOOKS------------");
            for (Books book : availableBooks) {
                System.out.printf("| TITLE: |" + book.getTitle() + "| NUMBER OF RENTS: | " +  book.getRentCounter());
                if(!book.isRented()){
                    System.out.print(" | Without Owner | ");
                }
                else{
                    System.out.printf(" | OWNER: | " + book.getOwner().getName());
                }
            }
        }
}
}


